Userland/Libraries/LibCrypto/Checksum/CRC32.cpp

```diff
@@ -1,17 +1,38 @@
 /*
- * Copyright (c) 2020, the SerenityOS developers.
+ * Copyright (c) 2020-2022, the SerenityOS developers.
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/Array.h>
 #include <AK/Span.h>
 #include <AK/Types.h>
 #include <LibCrypto/Checksum/CRC32.h>
 
 namespace Crypto::Checksum {
 
+static constexpr auto generate_table()
+{
+    Array<u32, 256> data {};
+    for (auto i = 0u; i < data.size(); i++) {
+        u32 value = i;
+
+        for (auto j = 0; j < 8; j++) {
+            if (value & 1) {
+                value = 0xEDB88320 ^ (value >> 1);
+            } else {
+                value = value >> 1;
+            }
+        }
+
+        data[i] = value;
+    }
+    return data;
+}
+
 void CRC32::update(ReadonlyBytes data)
 {
+    constexpr auto table = generate_table();
```

Comment:
:thinking: won't this have to be copied onto the stack now?

Reply:
That does seem to be what it's doing:
https://serenity.godbolt.org/z/oEYzbTG7j

```asm
update(AK::Span<unsigned char const>):
        sub     rsp, 912
        mov     r8, rsi
        mov     rdx, rdi
        mov     esi, OFFSET FLAT:.LC0 ;; note: table data
        lea     rdi, [rsp-120]
        mov     ecx, 128
        rep movsq
        test    r8, r8
        je      .L1
        mov     eax, DWORD PTR state[rip]
        mov     rdi, rdx
        lea     rsi, [r8+rdx]
.L3:
        movzx   edx, BYTE PTR [rdi]
        add     rdi, 1
        xor     edx, eax
        shr     eax, 8
        movzx   edx, dl
        xor     eax, DWORD PTR [rsp-120+rdx*4]
        mov     DWORD PTR state[rip], eax
        cmp     rsi, rdi
        jne     .L3
.L1:
        add     rsp, 912
        ret
```

vs as an object with static duration:
https://serenity.godbolt.org/z/r18Was8b5

```asm
update(AK::Span<unsigned char const>):
        test    rsi, rsi
        je      .L1
        mov     eax, DWORD PTR state[rip]
        add     rsi, rdi
.L3:
        movzx   edx, BYTE PTR [rdi]
        add     rdi, 1
        xor     edx, eax
        shr     eax, 8
        movzx   edx, dl
        xor     eax, DWORD PTR table[0+rdx*4] ;; note: direct address
        mov     DWORD PTR state[rip], eax
        cmp     rsi, rdi
        jne     .L3
.L1:
        ret
```

Reply:
Really interesting. I thought it the compiler would do a better job at constant propagation. I guess not. Here is a quick-bench that also shows the difference, quantitatively: https://quick-bench.com/q/zbWaG0jnEDfoiaA5Y3dcAFsGx8I

It looks like if the size of the CRC32 table was 64, it would propagate. https://serenity.godbolt.org/z/vnrTWc17T

Unfortunately, quick-bench doesn't seem to generate the SSE instructions so I can't send a quick link to the change there.

I will change it back to being a static, but keep the change to make it not a class and move it into the `.cpp`.

Reply:
> It looks like if the size of the CRC32 table was 64, it would propagate. [serenity.godbolt.org/z/vnrTWc17T](https://serenity.godbolt.org/z/vnrTWc17T)

that's not constprop, that's just copying the data on the stack with vector insns, you can see it still access the data off the stack:
```asm
xor     eax, DWORD PTR [rsp-120+rdx*4]
```

I doubt it would do any constprop for more than one value, unless the loop was force-unrolled somehow.